Document content of log.settings.options.txt (#4585)
mattermost/docs: source/administration/config-settings.rst

Output logs to multiple targets
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Allow any combination of console, local file, syslog, and TCP socket targets, and send log records to multiple targets. These targets have been chosen to support the vast majority of log aggregators, and other log analysis tools, without having to install additional software.

System Admins can define multiple log targets to:

- Mirror log output to files and log aggregators for redundancy.
- Log certain entries to specific destinations. For example, all errors could be routed to a specific destination for review.

Additional configuration options include:

- Multiple local file targets: Supports rotation and compression triggered by size and/or duration.
- Multiple syslogs: Supports local and remote syslog servers, with or without TLS transport.
- Multiple TCP sockets: TCP socket target can be configured with an IP address or domain name, port, and optional TLS certificate.

> All access to the REST API or CLI is audited. System Admins can configure whether the content of posts is included and whether permissions failures are audited.

Author: @wiggin77 - Can you tell me what config setting a System Admin would enable for this functionality, please? I'd like to include a link in the product documentation to the config setting.

Reviewer: When using the AdvancedLogging config option for audit and specifying a file or JSON config, an admin can add one or more targets (log destinations) and decide what discrete log items to include in each target.  To capture everything they would use 
```
"Levels": [
    {"ID": 100, "Name": "audit-api"},
    {"ID": 101, "Name": "audit-content"},
    {"ID": 102, "Name": "audit-permissions"},
    {"ID": 103, "Name": "audit-cli"},
],
```
in the target config.   NOTE, I made a mistake in my previous answer.  `audit-content` does not include the content of posts.  That level is there to allow admins to turn off audit of every post/reaction as it can be a fire-hose for busy servers.  However many customers would need that enabled.

Reviewer: `audit-api` - enables output of Rest API calls
`audit-content` - enables output of API calls that generate content (e.g. create post, create reaction)
`audit-permissions`- enables output of all permissions failures
`audit-cli` - enables output of legacy CLI calls

Author: Thanks @wiggin77! I've incorporated these details into this docs draft. Please review to ensure that I've captured the details accurately.

| Cert          | For TLS connections where TLS is set to ``true``, the filename of client certificate or base64-encoded certificate.                                    | string      |
+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+
|               |                                                                                                                                                        |             |
| Insecure      | Used for testing purposes only. Set to ``true`` to prevent a certificate check from being performed. Set to ``false`` to perform a certificate check.  | bool        |
+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+

.. Note::
    Filenames for ``AdvancedLoggingConfig`` can be absolute or relative, and can also contain the JSON directly embedded.

> See `Advanced Logging Options Sample.json.zip <https://docs.mattermost.com/source/samples/Advanced.Logging.Options.Sample.json.zip>`__ for a sample configuration file. 

Author: ```suggestion
See `Advanced Logging Options Sample.json.zip <https://github.com/mattermost/docs/blob/master/source/samples/Advanced.Logging.Options.Sample.json.zip>`__ for a sample configuration file. 
```

> See `Advanced Logging Options Sample.json.zip <https://github.com/mattermost/docs/files/5066597/Advanced.Logging.Options.Sample.json.zip>`__ for a sample configuration file. 

Author: ```suggestion
See `Advanced Logging Options Sample.json.zip <https://github.com/mattermost/docs/tree/master/source/samples/Advanced.Logging.Options.Sample.json.zip>`__ for a sample configuration file. 
```

> See `Advanced Logging Options Sample.json.zip <https://github.com/mattermost/docs/source/samples/Advanced.Logging.Options.Sample.json.zip>`__ for a sample configuration file. 

Author: ```suggestion
See `Advanced Logging Options Sample.json.zip <https://docs.mattermost.com/source/samples/Advanced.Logging.Options.Sample.json.zip>`__ for a sample configuration file. 
```

> See `Advanced Logging Options Sample.json.zip <https://github.com/mattermost/docs/source/samples/advanced-logging-options-sample-json.zip>`__ for a sample configuration file. 

Author: ```suggestion
See the `Advanced Logging Options Sample JSON ZIP file <https://github.com/mattermost/docs/source/samples/advanced-logging-options-sample-json.zip>`__ for a sample configuration file. 
```

> See `Advanced Logging Options Sample.json.zip <https://github.com/mattermost/docs/tree/master/source/samples/Advanced.Logging.Options.Sample.json.zip>`__ for a sample configuration file. 

Author: ```suggestion
See `Advanced Logging Options Sample.json.zip <https://github.com/mattermost/docs/source/samples/Advanced.Logging.Options.Sample.json.zip>`__ for a sample configuration file. 
```

> See `Advanced Logging Options Sample.json.zip <https://github.com/mattermost/docs/blob/master/source/samples/Advanced.Logging.Options.Sample.json.zip>`__ for a sample configuration file. 

Author: ```suggestion
See `Advanced Logging Options Sample.json.zip <https://github.com/mattermost/docs/blob/master/source/samples/advanced-logging-options-sample.json.zip>`__ for a sample configuration file. 
```

> See `Advanced Logging Options Sample.json.zip <https://github.com/mattermost/docs/blob/master/source/samples/advanced-logging-options-sample.json.zip>`__ for a sample configuration file. 

Author: Note to self: This download link isn't working as expected. Currently troubleshooting to identify fix.

Author: Download link now working with a download directive and a relative file path.

- Multiple local file targets: Supports rotation and compression triggered by size and/or duration.
- Multiple syslogs: Supports local and remote syslog servers, with or without TLS transport.
- Multiple TCP sockets: TCP socket target can be configured with an IP address or domain name, port, and optional TLS certificate.

All access to the REST API or CLI is audited. When using Advanced Logging for auditing, System Admins can capture the following auditing in the target configuration in addition to discrete log levels:

.. code-block:: none

   "Levels": [
      {"ID": 100, "Name": "audit-api"},
      {"ID": 101, "Name": "audit-content"},
      {"ID": 102, "Name": "audit-permissions"},
      {"ID": 103, "Name": "audit-cli"},
   ],

Where:

- ``audit-api``: Enables output of REST API calls.
- ``audit-content``: Enables output of API calls that generate content (e.g. ``create post``, ``create reaction``).
- ``audit-permissions``: Enables output of all permissions failures.
- ``audit-cli``: Enables output of legacy CLI calls.

.. Note::
  - Logs are recorded asynchronously to reduce latency to the caller. 
  - Advanced logging supports hot-reloading of logger configuration.

> +----------------------------------------------------------------------------------------------------------------------------------------------------------------------+

Reviewer 2: @cwarnermm This table is not rendering in the docs https://docs.mattermost.com/administration/config-settings.html#output-logs-to-multiple-targets

Output logs to multiple targets
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Allow any combination of console, local file, syslog, and TCP socket targets, and send log records to multiple targets. These targets have been chosen to support the vast majority of log aggregators, and other log analysis tools, without having to install additional software.

Reviewer 3: ```suggestion
Allow any combination of console, local file, syslog, and TCP socket targets, and send log records to multiple targets. These targets have been chosen as they support the vast majority of log aggregators, and other log analysis tools, without needing additional software installed.
```

System Admins can define multiple log targets to:

- Mirror log output to files and log aggregators for redundancy.
- Log certain entries to specific destinations. For example, all errors could be routed to a specific destination for review.

Additional configuration options include:

- Multiple local file targets: Supports rotation and compression triggered by size and/or duration.
- Multiple syslogs: Supports local and remote syslog servers, with or without TLS transport.
- Multiple TCP sockets: TCP socket target can be configured with an IP address or domain name, port, and optional TLS certificate.

> All access to the REST API or CLI is audited. When using Advanced Logging for audit and specifying a file or JSON configuration, System Admins can capture the following auditing in the target configuration:

Reviewer: Not restricted to file target or JSON formatter.  Any target can be used for auditing and the discrete log levels can be used.

.. code-block:: none

   "Levels": [
      {"ID": 100, "Name": "audit-api"},
      {"ID": 101, "Name": "audit-content"},
      {"ID": 102, "Name": "audit-permissions"},
      {"ID": 103, "Name": "audit-cli"},
   ],

Where:

> - ``audit-api `` - Enables output of REST API calls.

Author: ```suggestion
- ``audit-api`` - Enables output of REST API calls.
```

- ``audit-content`` - Enables output of API calls that generate content (e.g. create post, create reaction).
- ``audit-permissions`` - Enables output of all permissions failures.
- ``audit-cli`` - Enables output of legacy CLI calls.

> Beyond the standard log levels (trace, debug, info, panic), discrete log levels can also be specified. New log levels must be registered in ``mattermost/mattermost-server/shared/mlog/levels.go``.

Reviewer: Adding new log levels is a developer issue and IMO shouldn't appear in end-user documentation.

Author: Good point, @wiggin77. Can a customer with developer knowledge register new log levels within ``mattermost/mattermost-server/shared/mlog/levels.go``? Or is this something that Mattermost would complete on a customer's behalf?

Reviewer 4: I agree to remove this.  Customers can fork or add PRs if they require new log levels.

> .. Note::

Reviewer 4: Would this note also need to be removed as per the last comment?

Author: Yes. I've removed details about adding new log levels via a go file, but left in details about the fact that logs are recorded asynchronously, and that advanced logging supports hot-reloading.

| Stacktrace    | Set to ``true`` to generate a stacktrace. Set to ``false`` to prevent a stacktrace from being generated.                                               | bool        |
+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+
|               |                                                                                                                                                        |             |
| **Targets**   |                                                                                                                                                        |             |
+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+
|               |                                                                                                                                                        |             |

> | Type          | Can be one of: "console", "file", "syslog", or "tcp".                                                                                                  | string      |

Reviewer 3: ```suggestion
| Type          | Can be one of: ``console``, ``file``, ``syslog`` or ``tcp``.                                                                                                  | string      |
```

+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+
|               |                                                                                                                                                        |             |

> | Format        | Can be either "json" or "plain".                                                                                                                       | string      |

Reviewer 3: ```suggestion
| Format        | Can be either ``json`` or ``plain``.                                                                                                                       | string      |
```

| Levels        | Array of log levels.                                                                                                                                   | []          |
+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+
|               |                                                                                                                                                        |             |
| Options       | Map of options specific to the target type.                                                                                                            | {}          |
+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+
|               |                                                                                                                                                        |             |

> | MaxQueueSize  | The number of audit records that can be queued/buffered at any poitn in time when writing to syslog. Default is 1000.                                  | int         |

Reviewer 3: ```suggestion
| MaxQueueSize  | The number of audit records that can be queued/buffered at any point in time when writing to syslog. Default is 1000.                                  | int         |
```

+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+
|               |                                                                                                                                                        |             |
| **Console**   |                                                                                                                                                        |             |
+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+
|               |                                                                                                                                                        |             |

> | Out           | Can be either "stdout" or "stderr".                                                                                                                    | string      |

Reviewer 3: ```suggestion
| Out           | Can be either ``stdout`` or ``stderr``.                                                                                                                    | string      |
```

| **File**      |                                                                                                                                                        |             |
+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+
|               |                                                                                                                                                        |             |
| Filename      | Path and filename for logs.                                                                                                                            | string      |
+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+
|               |                                                                                                                                                        |             |

> | MaxAgeDays    | Number of days until a rotation is triggered. Set to ``0 ``to not rotate based on age.                                                                 | int         |

Reviewer 3: ```suggestion
| MaxAgeDays    | Number of days until a rotation is triggered. Set to ``0`` to not rotate based on age.                                                                 | int         |
```

|               |                                                                                                                                                        |             |
| Cert          | For TLS connections where TLS is set to ``true``, the filename of client certificate or base64-encoded certificate.                                    | string      |
+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+
|               |                                                                                                                                                        |             |
| Insecure      | Used for testing purposes only. Set to ``true`` to prevent a certificate check from being performed. Set to ``false`` to perform a certificate check.  | bool        |
+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+

.. Note::

>     Filenames for ``AdvancedLoggingConfig`` can be absolute or relative, and can also contain the JSON embedded directly.

Reviewer: I would word this something like, "AdvancedLoggingConfig can contain a absolute or relative filename, or embedded JSON."

> See the `Advanced Logging Options Sample JSON ZIP file <https://github.com/mattermost/docs/source/samples/advanced-logging-options-sample-json.zip>`__ for a sample configuration file. 

Author: ```suggestion
See the :download:`Advanced Logging Options Sample JSON ZIP file <https://github.com/mattermost/docs/source/samples/advanced-logging-options-sample-json.zip>`__ for a sample configuration file. 
```

> See the :download:`Advanced Logging Options Sample JSON ZIP file <https://github.com/mattermost/docs/source/samples/advanced-logging-options-sample-json.zip>`__ for a sample configuration file. 

Author: ```suggestion
See the :download:`Advanced Logging Options Sample JSON ZIP file <../samples/advanced-logging-options-sample-json.zip>` for a sample configuration file. 
```
